### ISSUE TITLE
fincore cannot handle large files

### ISSUE THREAD

**Issue by pkolano** [hirose31/Sys-PageCache#1]
I was trying out your Sys::PageCache module and noticed that the fincore method can't handle large files (e.g. 64GB) because it tries to map the whole file into memory.  BTW, the linux-ftools project mentioned in your "see also" section has the same issue.  To fix, you have to break the file up into chunks and mmap/mincore each chunk individually. You can see a patch someone submitted for linux-ftools here:

    https://code.google.com/archive/p/linux-ftools/issues/7

The same approach should be usable in your module.  Thanks!

**Comment by hirose31** [comment]
Could you show me `ulimit -v` and try again with `ulimit -v unlimited`?

```
$ truncate -s 10G file10G

$ sudo  bash -c 'ulimit -v; pagecache-check.pl file10G'
524288
mmap: Cannot allocate memory at /.../Sys/PageCache.pm line 64. at /usr/bin/pagecache-check.pl line 30.
Use of uninitialized value in printf at /usr/bin/pagecache-check.pl line 32.
Use of uninitialized value in printf at /usr/bin/pagecache-check.pl line 32.
Use of uninitialized value in printf at /usr/bin/pagecache-check.pl line 32.
Use of uninitialized value in printf at /usr/bin/pagecache-check.pl line 32.
file10G     : cached/total_size=0/0 cached/total_pages=0/0


$ sudo  bash -c 'ulimit -v unlimited; ulimit -v; pagecache-check.pl file10G'
unlimited
file10G     : cached/total_size=0/10737418240 cached/total_pages=0/2621440
```

`Sys::PageCache#fincore` does not allocate any memory because it calls `mmap` with `PROT_NONE` (Pages may not be accessed). But it seems the resource limit has effect.

**Comment by pkolano** [comment]
yes, you are right.  If you use "ulimit -v unlimited", it does work
correctly (the default "ulimit -v" is 54306080 on my system).  I tried on
an 850 GB file with no issues so it must indeed not be allocating memory
like you said.  While this is ok for a workaround, the problem is that you
can't guarantee the ulimit settings of someone invoking a program that
calls your module.  A partial solution might be invoking setrlimit
appropriately before mmap.  There would still be an issue, however, if
there was an admin-specified hard limit on the system that could not be
modified by the user.  In that case, it seems like the only full fix is
mmap/mincore in chunks of at most the soft limit (or hard limit if soft is
set to hard via setrlimit).

**Comment by hirose31** [comment]
I've just released v0.04. It will be available soon.

In this release Sys::PageCache#fincore calls mmap(2) in chunks (default max size is 512MB).
Could you try v0.04?

**Comment by pkolano** [comment]
The new version works great.  Thank you.
